### PR TITLE
Add exercise.parse.checker option for checking parse errors

### PR DESCRIPTION
**Pull request by cpsievert**
Related to https://github.com/rstudio/learnr/issues/356 (and https://github.com/rstudio/learnr/pull/403)

### CODE REVIEW

**Comment by cpsievert** [comment]
@schloerke is https://github.com/rstudio-education/gradethis/issues/109 referring to a different error (that a parse/eval error in the user submitted code)?

**Comment by schloerke** [comment]
Correct.

I believe I was referring to a parse error or `learnr` returning an error condition (eval error) in `last_value` or something.

File: R/grade_learnr.R [hunk 1]
```diff
@@ -44,27 +44,49 @@ grade_learnr <- function(label = NULL,
...
                          ...) {
-
-  # Sometimes no user code is provided, but
-  # that means there is nothing to check. Also,
-  # you do not want to parse NULL
-  if (is.null(user_code)) {
-    return(list(
+  
+  # copy in all learnr arguments
+  learnr_args <- list(...)
```
Comment (wch) [inline R/grade_learnr.R:49]: You should be able to just do:

```R
 learnr_args <- list(
    ...,
    label = label,
    solution_code = solution_code
  )
```

and so on

File: R/grade_learnr.R [hunk 2]
```diff
@@ -44,27 +44,49 @@ grade_learnr <- function(label = NULL,
...
+    evaluate_result = evaluate_result,
+    envir_prep = envir_prep,
+    last_value = last_value
+  )
+  
+  user_code <- tryCatch(
+    parse(text = user_code %||% ""),
+    error = function(e) {
+      parse_checker <- getOption(
+        "exercise.parse.error",
```
Comment (cpsievert) [inline R/grade_learnr.R:63]:
```suggestion
        "gradethis.parse.error", 
```

File: R/grade_learnr.R [hunk 3]
```diff
@@ -153,12 +165,22 @@ grade_learnr <- function(label = NULL,
...
+  feedback(
     message = checked_result$message,
     correct = checked_result$correct,
     type = message_type,
     location = "append"
   )
+}
+
+
+feedback <- function(message, correct, type, location) {
```
Comment (cpsievert) [inline R/grade_learnr.R:177]: Export this and `is_feedback()`

Comment (cpsievert) [reply R/grade_learnr.R:177]: Add checkmate checks for all the args

File: inst/tutorials/grading-demo/grading-demo.Rmd [hunk 4]
```diff
@@ -14,7 +14,10 @@ description: "Learn how to use gradethis to provide automated, personalized, for
 library(learnr)
 library(gradethis)
 
-tutorial_options(exercise.timelimit = 60, exercise.checker = gradethis::grade_learnr)
+tutorial_options(
+  exercise.timelimit = 60, 
+  exercise.checker = grade_learnr
+)
```
Comment (cpsievert) [inline inst/tutorials/grading-demo/grading-demo.Rmd:20]: Use new `exercise.error.checker = grade_error`

**Comment by cpsievert** [comment]
To be merged after #127

File: R/grade_learnr.R (hunk 3 above)
Comment (cpsievert) [reply R/grade_learnr.R]: I ended up calling this `grade_feedback()` which could supercede `graded()` (https://github.com/rstudio-education/gradethis/pull/123/commits/c449343cbc8696a280b45728604b4213c36e6129). 

@schloerke, do you have an argument for not binding the `graded()` object more closely to learnr feedback? (to me, in this case, it doesn't seem worth decoupling the grade from feedback)

Comment (cpsievert) [reply R/grade_learnr.R]: After discussion, we're thinking `graded()` should not be deprecated, and `grade_feedback` should have the implementation:

```r
grade_feedback <- function(grade, type, location) {
  if (!is_not_grade(grade)) error
  list(
    message = grade$message,
    correct = grade$correct,
    type = type
    location = location
  )
}
}


```

File: DESCRIPTION [hunk 5]
```diff
@@ -36,7 +36,8 @@ Imports:
     checkmate,
     testthat
 Suggests:
-    tidyr
+    tidyr,
+    learnr
```
Comment (cpsievert) [inline DESCRIPTION:40]: Move to an import

File: R/grade_result.R [hunk 6]
```diff
@@ -59,17 +77,104 @@ grade_result <- function(
...
+      .incorrect = incorrect
+    )
   )
+}
 
+
+#' @rdname grade_result
+#' @export
+#' @inheritParams grade_code
+grade_result_strict <- function(
```
Comment (cpsievert) [inline R/grade_result.R:97]: TODO: think a little more about naming here

File: R/grade_result.R [hunk 7]
```diff
@@ -1,28 +1,46 @@
...
+#'   fail_if(~ !is.integer(.result), "I expected an integer"),
+#'   learnr_args = list(last_value = 5)
+#' )
+#' 
+#' grade_result(
+#'   fail_if(~ !is.function(.result), "I expected a function."),
+#'   fail_if(~ .result(1) != 2, "Your function should add one."),
+#'   learnr_args = list(last_value = function(x) x + 2)
+#' )
+#'
```
Comment (cpsievert) [inline R/grade_result.R:43]:
```suggestion
#' 
#' \dontrun{gradethis_demo()}
```

Comment (cpsievert) [reply R/grade_result.R:43]: Done in 106310da0fddda6c9b93b8a055b05203e3d56775

**Pull request merged**